platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +90,196 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
+            ServerErrorException.class);
       }
     }
 
-    if (request.getType() == GetRateRequest.Type.FIRM) {
-      if (rate.getId() == null || rate.getExpiresAt() == null) {
-        Log.error("'id' and/or 'expires_at' are missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+    // sell_amount is present and positive number
+    if (!isPositiveNumber(Objects.requireNonNull(rate).getSellAmount())) {
+      logErrorAndThrow(
+          "'rate.sell_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // buy_amount is present and positive number
+    if (!isPositiveNumber(rate.getBuyAmount())) {
+      logErrorAndThrow(
+          "'rate.buy_amount' is missing or not a positive number in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // sell_amount has a proper number of significant decimals
+    if (!hasProperSignificantDecimals(
+        Objects.requireNonNull(rate).getSellAmount(), sellAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          format(
+              "'rate.sell_amount' (%s) has incorrect number of significant decimals (expected: %d) in the GET /rate response",
+              Objects.requireNonNull(rate).getSellAmount(), sellAsset.getSignificantDecimals()),
+          ServerErrorException.class);
+    }
+
+    // buy_amount has a proper number of significant decimals
+    if (!hasProperSignificantDecimals(
+        Objects.requireNonNull(rate).getBuyAmount(), buyAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          format(
+              "'rate.buy_amount' (%s) has incorrect number of significant decimals (expected: %d) in the GET /rate response",
+              rate.getBuyAmount(), buyAsset.getSignificantDecimals()),
+          ServerErrorException.class);
+    }
+
+    FeeDetails fee = rate.getFee();
+    // if fee is set, check the following
+    if (fee != null) {
+      // fee.total is present and is a positive number
+      if (!isPositiveNumber(fee.getTotal())) {
+        logErrorAndThrow(
+            "'rate.fee.total' is missing or not a positive number in the GET /rate response",
+            ServerErrorException.class);
+      }
+      // fee.asset is a valid asset
+      AssetInfo feeAsset = assetService.getAssetByName(fee.getAsset());
+      if (fee.getAsset() == null || feeAsset == null) {
+        logErrorAndThrow(
+            "'rate.fee.asset' is missing or not a valid asset in the GET /rate response",
+            ServerErrorException.class);
+      }
+
+      if (!isPositiveNumber(Objects.requireNonNull(feeAsset.getSignificantDecimals()).toString())) {
+        logErrorAndThrow(
+            format(
+                "The rate fee.asset (%s) does not have proper significant decimals defined in the assets configuration.",
```

Comment:
Error message doesn't match the condition

Reply:
Fixed.